curp/src/client_new/retry.rs

```diff
@@ -0,0 +1,256 @@
+    /// Create a retry client
+    fn new(inner: Api, config: RetryConfig) -> Self {
+        Self { inner, config }
+    }
+
+    /// Takes a function f and run retry.
+    async fn retry<'a, R, F>(&'a self, f: impl Fn(&'a Api) -> F) -> Result<R, tonic::Status>
+    where
+        F: Future<Output = Result<R, CurpError>>,
+    {
+        let mut backoff = self.config.init_backoff();
+        while let Some(delay) = backoff.next_delay() {
+            let err = match f(&self.inner).await {
+                Ok(res) => return Ok(res),
+                Err(err) => err,
+            };
+
+            match err {
+                // some errors that should not retry
+                CurpError::Duplicated(_)
+                | CurpError::ShuttingDown(_)
+                | CurpError::InvalidConfig(_)
+                | CurpError::NodeNotExists(_)
+                | CurpError::NodeAlreadyExists(_)
+                | CurpError::LearnerNotCatchUp(_) => {
+                    return Err(tonic::Status::from(err));
+                }
+
+                // some errors that could have a retry
+                CurpError::ExpiredClientId(_)
+                | CurpError::KeyConflict(_)
+                | CurpError::RpcTransport(_)
+                | CurpError::Internal(_) => {}
+
+                // update the cluster state if got WrongClusterVersion
+                CurpError::WrongClusterVersion(_) => {
+                    // the inner client should automatically update cluster state when fetch_cluster
+                    if let Err(e) = self.inner.fetch_cluster(false).await {
+                        warn!("fetch cluster failed, error {e:?}");
+                    }
+                }
+
+                // update the leader state if got Redirect
+                CurpError::Redirect(Redirect { leader_id, term }) => {
+                    self.inner.update_leader(leader_id, term);
+                }
+            }
+
+            warn!("retry on {} seconds later", delay.as_secs_f32());
+            tokio::time::sleep(delay).await;
+        }
+
+        Err(tonic::Status::deadline_exceeded("request timeout"))
+    }
+}
+
+#[async_trait]
+impl<Api> ClientApi for Retry<Api>
+where
+    Api: ClientApi<Error = CurpError> + LeaderStateUpdate + Send + Sync + 'static,
+{
+    /// The client error
+    type Error = tonic::Status;
+
+    /// Inherit the command type
+    type Cmd = Api::Cmd;
+
+    /// Get the local connection when the client is on the server node.
+    fn local_connect(&self) -> Option<Arc<dyn ConnectApi>> {
+        self.inner.local_connect()
+    }
+
+    /// Send propose to the whole cluster, `use_fast_path` set to `false` to fallback into ordered
+    /// requests (event the requests are commutative).
+    async fn propose(
+        &self,
+        cmd: &Self::Cmd,
+        use_fast_path: bool,
+    ) -> Result<ProposeResponse<Self::Cmd>, tonic::Status> {
+        self.retry::<_, _>(|client| async move { client.propose(cmd, use_fast_path).await })
```

Comment:
It seems we could replace the async block with just `client.propose(cmd, use_fast_path)`.

Reply:
done